arangod/RestServer/SoftShutdownFeature.cpp

```diff
@@ -0,0 +1,234 @@
+
+namespace {
+
+bool queueShutdownChecker(std::mutex& mutex,
+    arangodb::Scheduler::WorkHandle& workItem,
+    std::function<void(bool)>& checkFunc) {
+
+  arangodb::Scheduler* scheduler = arangodb::SchedulerFeature::SCHEDULER;
+  bool queued = false;
+  std::lock_guard<std::mutex> guard(mutex);
+  std::tie(queued, workItem) =
+      scheduler->queueDelay(arangodb::RequestLane::CLUSTER_INTERNAL,
+                            std::chrono::seconds(2),
+                            checkFunc);
+  return queued;
+}
+
+}
+
+namespace arangodb {
+
+SoftShutdownFeature::SoftShutdownFeature(application_features::ApplicationServer& server)
+    : ApplicationFeature(server, "SoftShutdown") {
+  setOptional(true);
+  startsAfter<application_features::AgencyFeaturePhase>();
+  startsAfter<ShutdownFeature>();
+  startsAfter<ConsoleFeature>();
+  startsAfter<ScriptFeature>();
+
+  // We do not yet know if we are a coordinator, so just in case,
+  // create a SoftShutdownTracker, it will not hurt if it is not used:
+  _softShutdownTracker = std::make_shared<SoftShutdownTracker>(server);
+}
+
+void SoftShutdownFeature::beginShutdown() {
+  _softShutdownTracker->cancelChecker();
+}
+
+void SoftShutdownTracker::cancelChecker() {
+  if (_softShutdownOngoing.load(std::memory_order_relaxed)) {
+    // This is called when an actual shutdown happens. We then want to
+    // delete the WorkItem of the soft shutdown checker, such that the
+    // Scheduler does not have any cron jobs any more:
+    _workItem.reset();
+  }
+}
+
+SoftShutdownTracker::SoftShutdownTracker(
+    application_features::ApplicationServer& server)
+  : _server(server), _softShutdownOngoing(false) {
+  _checkFunc = [this](bool cancelled) {
+    if (_server.isStopping()) {
+      return;   // already stopping, do nothing, and in particular
+                // let's not schedule ourselves again!
+    }
+    if (!this->check()) {   // Initiates shutdown if counters are 0
+      // Rearm ourselves:
+      if (!::queueShutdownChecker(this->_workItemMutex, this->_workItem,
+                                  this->_checkFunc)) {
+        // If this does not work, shut down right away:
+        this->initiateActualShutdown();
+      }
+    }
+  };
+}
+
+void SoftShutdownTracker::initiateSoftShutdown() {
+  bool prev = _softShutdownOngoing.exchange(true, std::memory_order_relaxed);
+  if (prev) {
+    // Make behaviour idempotent!
+    LOG_TOPIC("cce32", INFO, Logger::STARTUP)
+        << "Received second soft shutdown request, ignoring it...";
+    return;
+  }
+
+  LOG_TOPIC("fedd2", INFO, Logger::STARTUP)
+      << "Initiating soft shutdown...";
+
+  // Tell asynchronous job manager:
+  auto& generalServerFeature = _server.getFeature<GeneralServerFeature>();
+  auto& jobManager = generalServerFeature.jobManager();
+  jobManager.initiateSoftShutdown();
+
+  // Tell Pregel subsystem:
+  auto& pregelFeature = _server.getFeature<pregel::PregelFeature>();
+  pregelFeature.initiateSoftShutdown();
```

Comment:
I would prefer to have a method `initiateSoftShutdown()` in the ApplicationServer, which would call a new virtual method `initiazeSoftShutdown()` on each ApplicationFeature.
The default implementation of this method would do nothing of course, but any feature could override it is with something useful.

Reply:
I will do that.

Reply:
Done.